experimental/NeuralNets/Lib.idr

```diff
@@ -0,0 +1,151 @@
+grad_CrossEntropy [] _ = []
+grad_CrossEntropy _ [] = []
+grad_CrossEntropy [] [] = []
+grad_CrossEntropy (x::xs) (y::ys) = -(y/x) + (1-y)/(1-x)::grad_CrossEntropy xs ys
+
+export
+adam_Optimizer : List Double -> List Double -> Double -> Double -> Double -> Double -> Double -> Double -> List Double
+adam_Optimizer [] _ _ _ _ _ _ _ =  []
+adam_Optimizer _ [] _ _ _ _ _ _ =  []
+adam_Optimizer (t::thetas) (gt::gts) alpha m_prev v_prev beta1 beta2 epsilon = do
+              let m_current = (beta1 * m_prev + (1-beta1) * gt)/ (1-beta1)
+              let v_current = (beta2 * v_prev + (1-beta2) *  (db.pow gt 2))/(1-(db.pow beta2 2))
+              let mhat_current = m_current /(1-beta1)
+              let vhat_current = v_current/(1-beta2)
+              (t - alpha * mhat_current / ((sqrt  vhat_current) + epsilon)):: adam_Optimizer thetas gts alpha m_current v_current beta1 beta2 epsilon
+
+export
+wordpairs : List String -> List (String,String)
+wordpairs wrds = makeWordPairs wrds
+
+export
+xss : List (String, String) -> List String
+xss wordpairs = map (\(a,b) => a ) wordpairs
+
+export
+yss : List (String,String) -> List String
+yss wordpairs = map (\(a,b) => b ) wordpairs
+
+
+export
+wordify : List String -> List String
+wordify [] = []
+wordify (fst::xss) =
+            let bn = words fst
+            in bn ++ wordify xss 
+
+
+export
+rawToDict : List String -> List (String,Nat)
+rawToDict [] = []
+rawToDict dataset = let some_var = wordify dataset
+                    in createDictionary some_var
+
+export
+rawToCleaned : List String -> List String
+rawToCleaned [] = []
+rawToCleaned (x::xs) = let some_var = removeStopWords x
+                       in some_var ++ rawToCleaned xs
+export
+fullNNLayer : List (List Double) -> List (List Double) -> List (List Double)
+fullNNLayer [] [] = []
+fullNNLayer (x::inputs) (y::weights) = (softmax_activation $ nn_Layer x y)::fullNNLayer inputs weights
+
+export
+loss_func : (List Double -> List Double -> List Double ) -> List (List Double) -> List (List Double) -> List (List Double)
+loss_func f [] _ = []
+loss_func f _ [] = []
+loss_func f [] [] = []
+loss_func f (p::preds) (t::targets) = f p t :: loss_func f preds targets
+
+export
+optimizer : ( List Double -> List Double -> Double -> Double -> Double -> Double -> Double -> Double -> List Double) -> List (List Double) -> List (List Double) -> Double -> Double -> Double -> Double -> Double -> Double -> List (List Double)
+optimizer f [] _ _ _ _ _ _ _ = []
+optimizer f _ [] _ _ _ _ _ _ = []
+optimizer f [] [] _ _ _ _ _ _= []
+optimizer f (ws::wss) (cs::costs) alpha  m v beta1 beta2 epsilon =  (f ws cs alpha m v beta1 beta2 epsilon) :: optimizer f wss costs alpha m  v beta1 beta2 epsilon
+
+export
+model : List (List Double) -> List (List Double) -> List (List Double) -> Double -> Double -> Double -> Double -> Int -> List (List Double )
+model [] _ _ _ _ _ _ _ = []
+model _ [] _ _ _ _ _ _ = []
+model _ _ [] _ _ _ _ _ = []
+model [] [] [] _ _ _ _ _= []
```

Comment:
I think that pattern `model [] [] [] ... = []` is already covered above, but please correct me if I'm wrong.